feat(linter): implement unicorn/no-useless-undefined

cblh opened the pull request:
Resolves #3870

Hey there, thought I'd give this a try as it's tagged "good first issue" :) Let me know if there's anything that needs to change

Another part of the code also needs modification:
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/no-useless-undefined.js#L95

I tried to modify it. If there are any issues with these modifications, please let me know.               
```
fn is_has_function_return_type(node: &AstNode, ctx: &LintContext<'_>) -> bool {
    let Some(parent_node) = ctx.nodes().parent_node(node.id()) else {
        return false;
    };
    match parent_node.kind() {
        AstKind::ArrowFunctionExpression(arrow_func_express) => {
            arrow_func_express.return_type.is_some()
        }
        AstKind::Function(func) => func.return_type.is_some(),
        _ => is_has_function_return_type(parent_node, ctx),
    }
}
```

cblh committed:
feat(linter): implement unicorn/no-useless-undefined

cblh committed:
chore(linter): keep code style

cblh committed:
chore(linter): fix clippy error

cblh committed:
chore(linter): fix fmt error

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+                let parent_node_kind = parent_node.kind();
+
+                // `return undefined`
+                if let AstKind::ReturnStatement(ret_stmt) = parent_node_kind {
```
Use a match statement here.

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+fn is_match_ignore_func_name(name: &str) -> bool {
+    COMPARE_FUNCTION_NAMES.contains(&name)
+        // `array.push(undefined)`
+        || name == "push"
```
Create a static set for these names. Also, consider sorting them and using a binary search for faster checks.

cblh replied:
I attempted to use phf::phf_set! like this file, but I'm confused why using set still requires binary search.
https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/rules/unicorn/prefer_add_event_listener.rs#L68

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+    OxcDiagnostic::warn(
+        "eslint-plugin-unicorn(no-useless-undefined): Do not use useless `undefined`.",
+    )
+    .with_help("Consider using `null` instead.")
```
```suggestion
    .with_help("Consider removing `undefined` or using `null` instead.")
```

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+        if member_expr.property.name.as_str() == "bind")
+}
+
+fn is_undefined(arg: &Argument) -> bool {
```
We already have an utility function that does this, consider using it.

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+        (r"const foo = () => undefined", options_ignore_arrow_function_body()),
+    ];
+
+    let fail = vec![(r"let foo = undefined;", None)];
```
Should we be checking for cases such as parameter initializers and assignment expressions?

cblh replied:
Yes it has been added.

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+    let options_ignore_arguments = || Some(serde_json::json!({ "checkArguments": false }));
+    let options_ignore_arrow_function_body =
+        || Some(serde_json::json!({"checkArrowFunctionBody": false}));
+    let pass = vec![
```
Can you please add cases for member expression assignment and class property assignment? These should all pass:
```ts
const x = { a: undefined }
const y: any = {}
y.foo = undefined

class Foo () {
  public x: number | undefined = undefined }
}

function foo(x: string | undefined = undefined) { } // maybe fix into `x?: string`?
```

cblh replied:
I will add cases for member expression assignment and class property assignment，but`function foo(x: string | undefined = undefined) { } maybe fix into `x?: string`?` should be fix like that: `x?: string`

I try add these case on:
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/test/no-useless-undefined.mjs

but get `?: string | undefined`

```
valid(2): function foo(x: string | undefined = undefined) { }
  Should have no errors but had 1: [
    {
      ruleId: 'rule-to-test/no-useless-undefined',
      severity: 1,
      message: 'Do not use useless `undefined`.',
      line: 1,
      column: 38,
      nodeType: 'Identifier',
      messageId: 'no-useless-undefined',
      endLine: 1,
      endColumn: 47,
      fix: { range: [Array], text: '?: string | undefined' }
    }
  ]
```

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+                                .semantic()
+                                .trivias()
+                                .comments_range(ret_stmt.span.start..ret_stmt.span.end);
+                            if let Some((_, comment)) = comments_range.last() {
```
This code can be simplified.

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,403 @@
+    }
+}
+
+fn is_function_bind_call(call_expr: &CallExpression) -> bool {
```
Consider utilizing `ast_utils::is_method_call`

cblh committed:
chore(linter): Improve no_useless_undefined.rs: Add `ast_util::is_method_call`, refine diagnostics, and handle more cases.

- Introduced `ast_util::is_method_call` to streamline method call checks.
- Enhanced diagnostics and suggested fixes for clearer code.
- Updated static set `FUNCTION_NAMES` to manage function names more efficiently.
- Extended handling for `return`, `yield`, arrow functions, and variable declarations using `undefined`.

cblh committed:
chore(linter): Refactor fixer logic in `no_useless_undefined.rs` for clarity and efficiency.

Simplified the deletion logic in `impl Rule for NoUselessUndefined` to handle `undefined` removal more efficiently.

cblh committed:
Merge branch 'main' into feature/20240704no-useless-undefined

cblh committed:
refactor(linter): optimize `no_useless_undefined` rule implementation

- Updated `no_useless_undefined` rule to handle cases where `undefined` is used in parameter initializers, assignment expressions, and return statements in functions with explicit return types.
- Added utility function `is_has_function_return_type` to check if a function has an explicit return type.
- Added multiple new test cases to cover scenarios including:
  - Object properties and class fields with `undefined`.
  - Functions and methods returning `undefined`.
  - Various valid uses of `undefined` that should not trigger the rule.
  - Improved handling of nested functions returning `undefined`.
- Improved existing test cases to ensure the rule correctly ignores cases where `undefined` is necessary.
- Added TODO comments for potential future enhancements.

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,670 @@
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUselessUndefined {
```
in from_confuguration, these default to `true`. However, `#[derive(Default)]` will automatically default booleans to false. Maybe implement Default manually?

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
```diff
@@ -0,0 +1,670 @@
+        ("function foo([bar = undefined] = []) {}", None),
+        ("foo(bar, undefined, undefined);", None),
+        ("let a = undefined, b = 2;", None),
+        // TODO: Handle parenthesized undefined
```
There's a method to get the inner expression for parenthesis; I don't remember the name off the top of my head but I'll find it for you.

cblh replied:
Thank you

jordan-boyer replied:
if you have an expression you can use the without_parenthesized() method =)

cblh committed:
feat(linter): set default values to true in `NoUselessUndefined` rule

Modified the `NoUselessUndefined` struct to set default values for `check_arguments` and `check_arrow_function_body` to `true`. This change ensures that these fields are initialized with `true` when using the `Default` trait.

- Removed `Default` derive macro from `NoUselessUndefined`.
- Implemented custom `Default` trait for `NoUselessUndefined` to set `check_arguments` and `check_arrow_function_body` to `true`.

cblh committed:
Merge branch 'main' into feature/20240704no-useless-undefined

cblh committed:
feat(linter): correct `no_useless_undefined` rule and adjust lint category

- Changed the category of `NoUselessUndefined` rule from `pedantic` to `suspicious` for better classification.
- Fixed `is_undefined` function to handle cases where the argument is not an expression.
- Ensured `is_undefined` function correctly checks for undefined identifiers in arguments.

cblh committed:
feature(linter): handle parenthesized `undefined` in `no_useless_undefined` rule

- Updated `no_useless_undefined.rs` to correctly traverse and handle parenthesized expressions containing `undefined`.
- Modified test cases to include scenarios with parenthesized `undefined`.
- Updated snapshot file to reflect changes in linting behavior and error messages.

Previously, the linter did not handle `undefined` wrapped in parentheses correctly, leading to missed linting opportunities. This update ensures that `undefined` within any number of nested parentheses is properly identified and linted.

cblh committed:
Merge branch 'main' into feature/20240704no-useless-undefined

cblh committed:
feat(linter): correct comment handling in `no_useless_undefined` rule

DonIsaac reviewed (approved):
LGTM: I ran the fixer and saw no parse errors on a second pass. I'd like another review from a second pair of eyes if possible, @mysteryven think you could give me a hand?

Boshen commented on crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs:
````diff
@@ -0,0 +1,687 @@
+    /// let foo;
+    /// ```
+    NoUselessUndefined,
+    correctness,
````
Should be `pedantic`?

cblh replied:
It should be `pedantic`, because checking function parameters will cause false positives in some functions. like：compare function names 、array function

cblh committed:
feat(linter): update rule severity for no_useless_undefined

mysteryven commented:
```js
test('extension is not outdated when there is no local and gallery', () => {
    const extension = instantiationService.createInstance(Extension, () => ExtensionState.Installed, () => undefined, undefined, undefined, undefined, undefined);
    assert.strictEqual(extension.outdated, false);
});
```

Our fixer need to run twice to fix this, Tested on `unicorn/no-useless-undefined`, it needs once.

Other parts are look good to me.

cblh commented:
> ```js
> test('extension is not outdated when there is no local and gallery', () => {
>     const extension = instantiationService.createInstance(Extension, () => ExtensionState.Installed, () => undefined, undefined, undefined, undefined, undefined);
>     assert.strictEqual(extension.outdated, false);
> });
> ```
> 
> https://github.com/microsoft/vscode/blob/c37f4a49e28831cec1aec1af0f626edaada1f7ce/src/vs/workbench/contrib/extensions/test/electron-sandbox/extension.test.ts#L29-L32
> 
> Our fixer need to run twice to fix this, tested on `unicorn/no-useless-undefined`, it needs once.
> 
> Other parts are also look good to me.

ok, I will fix it

cblh commented:
@mysteryven @DonIsaac 

```rust
            if i64::from(start) <= last_pos {
                return;
            }
```

https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/fixer.rs#L278

I found that `span` is an interval that is closed on the left and open on the right. It should not be `<=`, which will cause the continuous interval to be skipped. It should be `<`.

The test case needs to be modified to allow continuous interval modification, but I found that it will cause `no_useless_escape` test failure

https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/fixer.rs#L520
```rust
    const REMOVE_START: Fix = Fix::delete(Span::new(0, 4));
    const REPLACE_ID: Fix = Fix { span: Span::new(4, 10), content: Cow::Borrowed("foo") };
    fn apply_two_fix_when_the_start_the_same_as_the_previous_end() {
        let result = get_fix_result(vec![
            create_message(remove_start(), Some(REMOVE_START)),
            create_message(replace_id(), Some(REPLACE_ID)),
        ]);
        assert_eq!(result.fixed_code, TEST_CODE.replace("var answer", "foo"));
        assert!(result.fixed);
    }
```

mysteryven commented:
> I found that span is an interval that is closed on the left and open on the right. It should not be <=, which will cause the continuous interval to be skipped. It should be <.

Since it not the issue about current rule, let's merge this PR first.
